web/js/codeworld.js

```diff
@@ -892,15 +894,58 @@ function formatSource() {
     return;
   }
 
-  const src = window.codeworldEditor.getValue();
+  const src = doc.getValue();
   const data = new FormData();
   data.append('source', src);
   data.append('mode', window.buildMode);
 
+  const oldCursorPosition = doc.getCursor();
+  const lineContentAtOldCursorPosition = doc.getLine(oldCursorPosition.line);
+
   sendHttp('POST', 'indent', data, (request) => {
     if (request.status === 200) {
-      if (request.responseText !== src) {
-        codeworldEditor.getDoc().setValue(request.responseText);
+      const reformattedSrc = request.responseText;
+
+      if (reformattedSrc !== src) {
+        doc.setValue(reformattedSrc);
+
+        let oldLine = 0;
+        let newLine = 0;
+
+        Diff.diffLines(src, reformattedSrc).forEach((diff) => {
+          const { value, count, added, removed } = diff;
+
+          if (added) {
+            if (oldCursorPosition.line <= oldLine) {
+              doc.setCursor(newLine, oldCursorPosition.ch);
```

Comment:
I may not have been as clear as I could before.  Here's my proposal:

```
const target = doc.getCursor();

let oldLine = 0;
let newLine = 0;

for (const diff of Diff.diffLines(src, reformattedSrc)) {
  const { count, added, removed } = diff;

  if (removed) {
    if (oldLine + count > target.line) {
      break;
    }
    oldLine += count;
  } else if (added) {
    newLine += count;
  } else {
    if (oldLine + count >= target.line) {
      newLine += target.line - oldLine;
      break;
    }

    oldLine += count;
    newLine += count;
  }
}

doc.setCursor(newLine, target.ch);
```

That doesn't do anything about the column, but you can add the logic to set the right column after the loop.

This is probably going to be wrong sometimes, and require tweaking.  In particular, I think if there are consecutive additions and removals, the cursor will move to the beginning of that block.  That will be a little tricker to work out, but honestly, let's just get something in that doesn't scroll all the way back to the top, and worry about those details later.

Reply:
It does miss sometimes as you anticipated, but the logic in your solution is much clearer. I especially like breaking out of the loop when the position is found, and the fact that the line position is retained if the cursor is located outside of the formatted piece of code.

I can't think of a way to compute new column value this way, though. We could match the old line content each time new lines are added, but there could be multiple occurrences, how to determine where exactly this old string belongs to?

Reply:
Hmm.  One thing you could do is use `diffChars` instead of `diffLines`.

```
const target = doc.indexFromPos(doc.getCursor());

let oldIndex = 0;
let newIndex = 0;

for (const diff of Diff.diffChars(src, reformattedSrc)) {
  const { count, added, removed } = diff;

  if (removed) {
    if (oldIndex + count > target) {
      break;
    }
    oldIndex += count;
  } else if (added) {
    newIndex += count;
  } else {
    if (oldIndex + count >= target.line) {
      newIndex += target - oldIndex;
      break;
    }

    oldIndex += count;
    newIndex += count;
  }
}

doc.setCursor(doc.posFromIndex(newIndex));
```

I don't know if that would be too slow, but it probably does as close as possible to the right thing.